packages/rspack/src/util/bindingVersionCheck.ts

```diff
@@ -0,0 +1,142 @@
+import { readdirSync, readFileSync } from "fs";
+import path from "path";
+
+const NodePlatformArchToAbi: Record<
+	string,
+	Record<string, string | Record<string, string>>
+> = {
+	android: {
+		arm64: "",
+		arm: "eabi"
+	},
+	win32: {
+		x64: "msvc",
+		ia32: "msvc",
+		arm64: "msvc"
+	},
+	darwin: {
+		x64: "",
+		arm64: ""
+	},
+	freebsd: {
+		x64: ""
+	},
+	linux: {
+		x64: {
+			musl: "musl",
+			gnu: "gnu"
+		},
+		arm64: {
+			musl: "musl",
+			gnu: "gnu"
+		},
+		arm: "gnueabihf"
+	}
+};
+
+function isMusl() {
+	// For Node 10
+	if (!process.report || typeof process.report.getReport !== "function") {
+		try {
+			const lddPath = require("child_process")
+				.execSync("which ldd")
+				.toString()
+				.trim();
+			return readFileSync(lddPath, "utf8").includes("musl");
+		} catch (e) {
+			return true;
+		}
+	} else {
+		// @ts-expect-error getReport returns an object containing header object
+		const { glibcVersionRuntime } = process.report.getReport().header;
+		return !glibcVersionRuntime;
+	}
+}
+
+const BINDING_VERSION = require("@rspack/binding/package.json").version;
+const CORE_VERSION = require("../../package.json").version;
+
+const getAddonPlatformArchAbi = function () {
+	const { platform, arch } = process;
+	let binding = "";
+	binding += platform;
+
+	let abi = NodePlatformArchToAbi[platform][arch];
+	if (abi === undefined) return new Error(`unsupported cpu arch: ${arch}`);
+	binding += "-" + arch;
+
+	if (typeof abi === "string") {
+		binding += abi.length ? "-" + abi : "";
+	} else if (typeof abi === "object") {
+		binding += "-" + abi[isMusl() ? "musl" : "gnu"];
+	} else {
+		return new Error(`unsupported abi: ${abi}`);
+	}
+
+	return binding;
+};
+
+/**
+ * Error: version checked with error
+ * null: version checked without any error
+ * undefined: version to be checked
+ */
+let result: Error | undefined | null;
+
+/**
+ * Check if these version matches:
+ * `@rspack/core`, `@rspack/binding`, `@rspack/binding-<platform>-<arch>-<abi>`
+ */
+export const checkVersion = () => {
+	if (result !== undefined) {
+		return result;
+	}
+
+	let platformArchAbi = getAddonPlatformArchAbi();
+	if (platformArchAbi instanceof Error) {
+		return (result = platformArchAbi);
+	}
+
+	let ADDON_VERSION;
+	try {
+		const BINDING_PKG_DIR = path.dirname(
+			require.resolve("@rspack/binding/package.json")
+		);
+
+		const isLocal = readdirSync(BINDING_PKG_DIR).some(
+			item => item === `rspack.${platformArchAbi}.node`
+		);
+
+		if (isLocal) {
```

Comment:
do we really need local check, i thought monorepo already make it works like third_party node_modules

Reply:
> do we really need local check, i thought monorepo already make it works like third_party node_modules

This is for local testing, such as running an example in rspack repo.

Reply:
I know this scenario, but don't we already make sure all the version are same in monorepo?

Reply:
Of course, but then we have to figure out if we are running locally, or `@rspack/binding-xxx` cannot be found.